mri/reconstructors/self_calibrating.py

```diff
@@ -170,8 +204,7 @@ def reconstruct(self, kspace_data, optimization_alg='pogm', x_init=None,
                 method=self.smaps_gridding_method,
                 n_cpu=self.n_jobs
             )
-            self.extra_grad_args['Smaps'] = Smaps
-            self.initialize_gradient_op(**self.extra_grad_args)
+            self.set_smaps(self, Smaps)
```

Comment:
If we are anyway using setters and getters why not do it the right way? We can that way make Smaps as a property of class with the functions as defined by you as setter and getter... 
You can refer https://medium.com/@pranaygore/setters-and-getters-in-python-76b5473b3c83 or anything similar..

Reply:
I wouldn't describe that as "the right way". Even the non-canonical source you refer to describes it as an optional way, and the PEP8 does not specify how to handle getters and setters. I am myself not a big fan of this syntax because the user has no way to know if s.he is accessing the attribute correctly.

But for sure we can store the `Smaps` as an explicit class attribute instead of a value stored in `self.extra_grad_args,` but currently there is no other use for it, and we would have to store it in `self.extra_grad_args` anyway. Also some may prefer `setSmaps` instead of `set_smaps`, since it also matches the capital letter of the variable.

Reply:
> but currently there is no other use for it, and we would have to store it in self.extra_grad_args anyway. 

Do we have to save it anyway even if we have the attribute like self._smaps? Why cant we explictly call as follows:

`self.initialize_gradient_op(**self.extra_grad_args, Smaps=self.smaps)` 

@zaccharieramzi do you have any thoughts on the best way to have a setter and getter?

The idea here is that we want to handle the following cases (please add if I missed anything):

1) Reinitialize gradient operator when Fourier operator changes (smaps are the same)
2) Get new smaps (maybe through other methods), and initialize gradient operator with them
3) Maybe just reconstruct for different images, with the same Smaps.

Reply:
> Do we have to save it anyway even if we have the attribute like self._smaps? Why cant we explictly call as follows:
> 
> self.initialize_gradient_op(**self.extra_grad_args, Smaps=self.smaps)

It means we would have to systematically align the default value of `Smaps` in the `SelfCalibrationReconstructor` to the one in the gradient operator initializer. We avoid that by just not adding anything to the `self.extra_grad_args` variable when not needed. That won't necessarily be an issue in the future but that something to keep in mind.

More generally I don't see the point of letting the user change every piece of the reconstructor when it's computationally cheap to just instantiate another reconstructor. The idea of this PR was to:

1. Be able to give acquired/pre-computed Smaps to the reconstructor instead of letting the reconstructor compute them at least once.
2. Give to the user a proper way to retrieve the Smaps in the case they were computed by the reconstructor (to later re-run the script without having to compute them thanks to 1.)

Reply:
I don't necessarily see a point in having the getters or setters in one way or another in this particular case, so I would say let's go with what @Daval-G has implemented which does the job described [here](https://github.com/CEA-COSMIC/pysap-mri/pull/79#discussion_r375354520).